src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+    };
+
+    size_t mIndex = MAX_SIZE_T;
+    Type mType = Type::UNKNOWN;
+
+    VariablePtr mInitialisingVariable;
+    VariablePtr mVariable;
+
+    AnalyserInternalEquationWeakPtr mEquation;
+
+    explicit AnalyserInternalVariable(const VariablePtr &variable);
+
+    void setVariable(const VariablePtr &variable,
+                     bool checkInitialValue = true);
+
+    void makeVoi();
+    void makeState();
+};
+
+using AnalyserInternalVariablePtr = std::shared_ptr<AnalyserInternalVariable>;
+
+AnalyserInternalVariable::AnalyserInternalVariable(const VariablePtr &variable)
+{
+    setVariable(variable);
+}
+
+void AnalyserInternalVariable::setVariable(const VariablePtr &variable,
+                                           bool checkInitialValue)
+{
+    if (checkInitialValue && !variable->initialValue().empty()) {
+        // The variable has an initial value, so it can either be a constant or
+        // a state. By default, we consider it to be a constant and, if we find
+        // an ODE for that variable, we will know that it was actually a state.
+
+        mType = Type::CONSTANT;
+
+        mInitialisingVariable = variable;
+    }
+
+    mVariable = variable;
+}
+
+void AnalyserInternalVariable::makeVoi()
+{
+    mType = Type::VARIABLE_OF_INTEGRATION;
+}
+
+void AnalyserInternalVariable::makeState()
+{
+    if (mType == Type::UNKNOWN) {
+        mType = Type::SHOULD_BE_STATE;
+    } else if (mType == Type::CONSTANT) {
+        mType = Type::STATE;
+    }
+}
+
+#ifdef SWIG
+struct AnalyserEquation
+#else
+struct AnalyserInternalEquation: public std::enable_shared_from_this<AnalyserInternalEquation>
+#endif
+{
+    enum struct Type
+    {
+        UNKNOWN,
+        TRUE_CONSTANT,
+        VARIABLE_BASED_CONSTANT,
+        RATE,
+        ALGEBRAIC
+    };
+
+    size_t mOrder = MAX_SIZE_T;
+    Type mType = Type::UNKNOWN;
+
+    std::vector<AnalyserInternalEquationPtr> mDependencies;
+
+    AnalyserEquationAstPtr mAst;
+
+    std::vector<AnalyserInternalVariablePtr> mVariables;
+    std::vector<AnalyserInternalVariablePtr> mOdeVariables;
+
+    AnalyserInternalVariablePtr mVariable;
+    ComponentPtr mComponent = nullptr;
+
+    bool mComputedTrueConstant = true;
+    bool mComputedVariableBasedConstant = true;
+
+    bool mIsStateRateBased = false;
+
+    explicit AnalyserInternalEquation(const ComponentPtr &component);
+
+    void addVariable(const AnalyserInternalVariablePtr &variable);
+    void addOdeVariable(const AnalyserInternalVariablePtr &odeVariable);
+
+    static bool containsNonUnknownVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+    static bool containsNonConstantVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+
+    static bool knownVariable(const AnalyserInternalVariablePtr &variable);
+    static bool knownOdeVariable(const AnalyserInternalVariablePtr &odeVariable);
+
+    bool check(size_t & equationOrder, size_t & stateIndex, size_t & variableIndex);
```

Comment:
Is this different from `&equationOrder` etc?

Reply:
What do you mean? Are you talking about `& equationOrder` vs. `&equationOrder`, i.e. with and without space? If so, that's a Clang-Tidy thing. Nothing we can do about it, if I recall correctly.

Reply:
... really? All the other equations have no space?

Reply:
Which other equations? Are you talking about
```
bool AnalyserInternalEquation::check(size_t &equationOrder, size_t &stateIndex,
                                     size_t &variableIndex)
```

If so, indeed, there is no space, but it's really a Clang-Tidy thing. If you do `make format_code` / `ninja format_code` after having put/removed spaces around `&`, you will see that in the interface code it will add a space around `&` while in the implementation code it will only add a space before `&`.